### PR TITLE
Add SIP-21: Snap-defined timeouts

### PR DESCRIPTION
**Pull request by FrederikBolding**
This PR adds SIP 21 called Snap-defined timeouts.

### CODE REVIEW

File: SIPS/sip-21.md [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+title: Snap-defined timeouts
+status: Draft
+discussions-to: https://github.com/MetaMask/SIPs/discussions/128
+author: Frederik Bolding (@frederikbolding)
+created: 2024-01-12
+---
+
+## Abstract
+
+This SIP describes a way for Snaps to customize the request timeouts that constrain the Snap lifecycle. Effectively allowing the Snap to opt-in to slightly longer or shorter timeouts where needed.
```
Comment (Mrtenz) [inline SIPS/sip-21.md:12]: What's the use case for a Snap requesting shorter timeouts?

File: SIPS/sip-21.md [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+
+## Motivation
+
+Snaps that have to do CPU intensive computation are currently limited by the default `1` minute timeout, causing certain use-cases to be unsupportable by Snaps (e.g. computing ZK proofs). By letting Snaps customize the request timeout (within reason) for each type of handler that they expose, the Snap developer gets more control over the user experience while the Snap platform lifecycle requirements can remain strict.
```
Comment (Mrtenz) [inline SIPS/sip-21.md:16]:
```suggestion
Snaps that do CPU intensive computation are currently limited by the 1 minute timeout, causing certain use-cases to be unsupportable by Snaps (e.g., computing ZK proofs). By letting Snaps customize the request timeout (within reason) for each type of handler that they expose, the Snap developer gets more control over the user experience while the Snap platform lifecycle requirements can remain strict.
```

File: SIPS/sip-21.md [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+
+The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT",
+"SHOULD", "SHOULD NOT", "RECOMMENDED", "NOT RECOMMENDED", "MAY", and
+"OPTIONAL" written in uppercase in this document are to be interpreted as described in [RFC 2119](https://www.ietf.org/rfc/rfc2119.txt)
+
+### Snap Manifest
+
+This SIP specifies an addendum that MUST be applied to existing and future Snap handler permissions (`endowment:rpc`, `endowment:transaction-insight` etc.). The addendum is an optional caveat value called `maxRequestTime`.
+
+The value MUST be a valid integer between `5000` (5 seconds) and `180000` (3 minutes) and specifies the request timeout for the given handler in milliseconds. If no value is provided the default timeout of 1 minute (`30000` ms) MUST be used.
```
Comment (Mrtenz) [inline SIPS/sip-21.md:32]: Is 3 minutes enough for the ZK Snaps?

Comment (FrederikBolding) [reply SIPS/sip-21.md:32]: I think so, but we will get some feedback from snap devs before proceeding!

File: SIPS/sip-21.md (hunk 1 above)
Comment (FrederikBolding) [reply SIPS/sip-21.md:12]: If a snap dev is confident that their snap runs faster than the 1 minute timeout and therefore confident that if it takes longer than 15 seconds something is wrong I think it makes sense for them to be able to use that as the timeout.

If something is actually wrong the user would be notified much quicker, making the UX slightly better.

File: SIPS/sip-21.md (hunk 3 above)
Comment (Montoya) [reply SIPS/sip-21.md:32]: Yes my understanding is 3 minutes would be enough.

Comment (Montoya) [reply SIPS/sip-21.md:32]: From Rarimo: "3 minutes should be enough but we may want to do up to 5 minutes for more complex ZKPs in the future."

"Usually it takes about 40s on average hardware and up to 2 minutes on slow devices to generate a ZKP at https://robotornot.rarimo.com/" 

"We’ll have estimates for a bit more complex ZKPs in our new project in the next two weeks"

File: SIPS/sip-21.md [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
```
Comment (ihordiachenko) [inline SIPS/sip-21.md:32]:
```suggestion
The value MUST be a valid integer between `5000` (5 seconds) and `180000` (3 minutes) and specifies the request timeout for the given handler in milliseconds. If no value is provided the default timeout of 1 minute (`60000` ms) MUST be used.
```
Fixed typo

Comment (FrederikBolding) [reply SIPS/sip-21.md:32]: Thanks!

**Review by Mrtenz — APPROVED** [review]

**Review by hmalik88 — APPROVED** [review]

**Pull request merged**